### PR TITLE
feat(web): register / edit poller in platformtopology via poller form

### PR DESCRIPTION
**Pull request by jeremyjaouen**
## Description
This PR allow the Configuration Poller formulary to create / update informations of a poller into the platform_topology page

**Fixes** #MON-5869

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 19.04.x
- [ ] 19.10.x
- [ ] 20.04.x
- [x] 20.10.x (master)

<h2> How this pull request can be tested ? </h2>

- Create a Poller using the Configuration Poller Form.
- You should see it into platform_topology table.
- Edit this poller (IP, Name, parent remote server)
- the Poller should have been updated into platform_topology table.

## Checklist

- [x] I followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [x] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have made corresponding changes to the **documentation**.
- [ ] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: www/include/configuration/configServers/DB-Func.php [hunk 1]
```diff
@@ -1093,3 +1104,206 @@ function defineLocalPollerToDefault()
...
+
+    /**
+     * In the case of editing a poller freshly duplicated, it doesn't exist in platform_topology,
+     * so we need to create it instead of editing
+     */
+    if ($platform) {
+        $statement = $pearDB->prepare("
+            UPDATE platform_topology
+            SET address = :address,
+            name = :name,
```
Comment (sc979) [inline www/include/configuration/configServers/DB-Func.php:1222]:
```suggestion
            UPDATE platform_topology SET
            address = :address,
            name = :name,
```

File: www/include/configuration/configServers/DB-Func.php [hunk 2]
```diff
@@ -1093,3 +1104,206 @@ function defineLocalPollerToDefault()
...
+    $statement->bindValue(':name', $name, \PDO::PARAM_STR);
+    $statement->bindValue(':type', $type, \PDO::PARAM_STR);
+    $statement->bindValue(':parent', $parentId, \PDO::PARAM_INT);
+    $statement->bindValue(':serverId', $serverId, \PDO::PARAM_INT);
+    $statement->execute();
+}
+
+/**
+ * Check if a poller IP can be registered and display an error in form if it can't
+ * This ruleset avoid IP duplication in Poller formulary
```
Comment (sc979) [inline www/include/configuration/configServers/DB-Func.php:1244]:
```suggestion
 * This ruleset avoid IP duplication in Poller form
```

File: www/include/configuration/configServers/DB-Func.php [hunk 3]
```diff
@@ -1093,3 +1104,206 @@ function defineLocalPollerToDefault()
...
+    if ($isAlreadyInTopology) {
+        return false;
+    }
+
+    return true;
+}
+
+/**
+ * Check if a poller IP can be updated and display an error in form if it can't
+ * This ruleset avoid IP duplication in Poller formulary
```
Comment (sc979) [inline www/include/configuration/configServers/DB-Func.php:1267]:
```suggestion
 * This ruleset avoid IP duplication in Poller form
```

File: www/include/configuration/configServers/DB-Func.php [hunk 4]
```diff
@@ -1093,3 +1104,206 @@ function defineLocalPollerToDefault()
...
+    $statement = $pearDB->prepare("
+        SELECT `id`, `ns_ip_address` AS `address` FROM `nagios_server`
+        WHERE `ns_ip_address` = :address
+    ");
+    $statement->bindValue(':address', $serverIp, \PDO::PARAM_STR);
+    $statement->execute();
+    $platform = $statement->fetch(\PDO::FETCH_ASSOC);
+
+    /**
+     * check if the previously found platform is the platform we're editing
```
Comment (sc979) [inline www/include/configuration/configServers/DB-Func.php:1288]:
```suggestion
     * check if previously found platform is the platform we're editing
```

File: www/include/configuration/configServers/DB-Func.php [hunk 5]
```diff
@@ -1093,3 +1104,206 @@ function defineLocalPollerToDefault()
...
+     */
+    if ($platform) {
+        if ((int) $platform['id'] === $serverId) {
+            return true;
+        }
+        return false;
+    } else {
+        /**
+         * If nothing was found in nagios server check if it existing in platform topology
```
Comment (sc979) [inline www/include/configuration/configServers/DB-Func.php:1297]:
```suggestion
         * If nothing was found in nagios server check if it exists in platform topology
```

File: www/include/configuration/configServers/DB-Func.php [hunk 6]
```diff
@@ -1093,3 +1104,206 @@ function defineLocalPollerToDefault()
...
+         * e.g: a Central is 127.0.0.1 in NS but is display with his true IP in platform_topology
```
Comment (sc979) [inline www/include/configuration/configServers/DB-Func.php:1298]:
```suggestion
         * e.g: a Central is 127.0.0.1 in NS but is displayed with its true IP in platform_topology
```

File: www/include/configuration/configServers/DB-Func.php [hunk 7]
```diff
@@ -1093,3 +1104,206 @@ function defineLocalPollerToDefault()
...
+        $statement = $pearDB->prepare("SELECT * FROM `platform_topology` WHERE `address` = :address");
+        $statement->bindValue(':address', $serverIp, \PDO::PARAM_STR);
+        $statement->execute();
+        $platformInTopology = $statement->fetch(\PDO::FETCH_ASSOC);
+        if ($platformInTopology) {
+            return false;
+        }
+        return true;
+    }
+}
```
Comment (sc979) [inline www/include/configuration/configServers/DB-Func.php:1309]:
```suggestion
}

```

**Review by sc979 — DISMISSED** [review]

File: www/include/configuration/configServers/DB-Func.php [hunk 8]
```diff
@@ -1093,3 +1104,212 @@ function defineLocalPollerToDefault()
...
+
+    $serverIp = $pollerInformations[':ns_ip_address'];
+    $serverName = $pollerInformations[':name'];
+    $type = (int) $pollerInformations[':localhost'] == true ? 'central' : 'poller';
+
+    /**
+     * Prepare statement to get the Parent depending on Remote attachment or not.
+     */
+    if (isset($pollerInformations[':remote_id'])) {
+        $statement = $pearDB->prepare("SELECT * FROM `platform_topology` WHERE `server_id` = :remoteId");
```
Comment (kduret) [inline www/include/configuration/configServers/DB-Func.php:1126]:
```suggestion
        $statement = $pearDB->prepare("SELECT id FROM `platform_topology` WHERE `server_id` = :remoteId");
```

File: www/include/configuration/configServers/DB-Func.php [hunk 9]
```diff
@@ -1093,3 +1104,212 @@ function defineLocalPollerToDefault()
...
+        $statement->bindValue(':remoteId', (int) $pollerInformations[':remote_id'], \PDO::PARAM_INT);
+        $statement->execute();
+    } else {
+        $statement = $pearDB->query("SELECT * FROM `platform_topology` WHERE `type` = 'central'");
```
Comment (kduret) [inline www/include/configuration/configServers/DB-Func.php:1130]:
```suggestion
        $statement = $pearDB->query("SELECT id FROM `platform_topology` WHERE `type` = 'central'");
```

File: www/include/configuration/configServers/DB-Func.php [hunk 10]
```diff
@@ -1093,3 +1104,212 @@ function defineLocalPollerToDefault()
...
+    } else {
+        /**
+         * Prepare statement to get the Parent depending on Remote attachment or not.
+         */
+        if (!empty($pollerInformations[':remote_id'])) {
+            $statement = $pearDB->prepare("SELECT * FROM `platform_topology` WHERE `server_id` = :remoteId");
+            $statement->bindValue(':remoteId', (int) $pollerInformations[':remote_id'], \PDO::PARAM_INT);
+            $statement->execute();
+        } else {
+            $statement = $pearDB->query("SELECT * FROM `platform_topology` WHERE `type` = 'central'");
```
Comment (kduret) [inline www/include/configuration/configServers/DB-Func.php:1197]:
```suggestion
            $statement = $pearDB->query("SELECT id FROM `platform_topology` WHERE `type` = 'central'");
```

File: www/include/configuration/configServers/DB-Func.php [hunk 11]
```diff
@@ -1093,3 +1104,212 @@ function defineLocalPollerToDefault()
...
+    }
+
+    if ($type === 'central') {
+        $parentId = null;
+    } else {
+        /**
+         * Prepare statement to get the Parent depending on Remote attachment or not.
+         */
+        if (!empty($pollerInformations[':remote_id'])) {
+            $statement = $pearDB->prepare("SELECT * FROM `platform_topology` WHERE `server_id` = :remoteId");
```
Comment (kduret) [inline www/include/configuration/configServers/DB-Func.php:1193]:
```suggestion
            $statement = $pearDB->prepare("SELECT id FROM `platform_topology` WHERE `server_id` = :remoteId");
```

File: www/include/configuration/configServers/DB-Func.php [hunk 12]
```diff
@@ -650,6 +650,12 @@ function insertServer(array $data): int
     $poller = $result->fetch();
     $result->closeCursor();
 
+    try {
+        insertServerIntoPlatformTopology($retValue, (int) $poller['last_id']);
+    } catch (Exception $e) {
```
Comment (kduret) [inline www/include/configuration/configServers/DB-Func.php:655]: Do you want to catch all types of exception, or only `\PDOException` ?

Comment (jeremyjaouen) [reply www/include/configuration/configServers/DB-Func.php:655]: PDOException and Exception that could be thrown while resolving the parent relation:
```
        if (!empty($parent['id'])) {
            $parentId = (int) $parent['id'];
        } else {
            throw new \Exception(
                'Missing parent platform topology. Please register the parent first using the endpoint
                or the available script. For more details check the documentation'
            );
        }
```

**Review by kduret — APPROVED** [review]

**Review by sc979 — APPROVED** [review]

**Pull request merged**